blog/2023-06-29-automated-testing/index.mdx

```diff
@@ -0,0 +1,134 @@
+
+When it comes to Weaviate applications, an integration test could cover the interaction between the application code, the Weaviate server, and the inference provider (e.g. OpenAI, Cohere, Google PaLM) used via a [vectorizer](/developers/weaviate/modules/retriever-vectorizer-modules) or [generator](/developers/weaviate/modules/reader-generator-modules) module. With the advent of [embedded Weaviate](/developers/weaviate/installation/embedded), integration testing has become much easier, since Weaviate can be instantiated directly from the client code, without the need to set up and tear down a separate server.
+
+
+## Scoping the test
+
+In integration testing, it's important to delineate precisely what should be tested, in order to avoid writing tests for the _units_ of a system, and instead focus on their _integration_. In the case of a typical application built with Weaviate,
+* Search quality depends on the selected model of the [vectorization module](/developers/weaviate/modules/retriever-vectorizer-modules). This should not be the focus of integration testing. In particular, returned [distances](../distance-metrics-in-vector-search) are subject to slight changes.
+* Search itself is a core Weaviate functionality, that is tested by Weaviate's own test suite. Inserting a dummy object and searching for it would duplicate part of this test suite, and wouldn't prove anything about the integration between Weaviate and the application.
+* Integration testing should focus on automatically catching common issues including:
+  * Connection or authentication issues with the inference provider (and with Weaviate, if not using it embedded)
+  * Incomplete or incorrect data imports
+  * Using the correct number of dimensions when [bringing your own vectors](/developers/weaviate/tutorials/custom-vectors)
+  * Schema issues, like invalid class names, properties, or data types
+  * Race conditions or concurrency errors from multiple requests
+
+
+## Sample integration test
+
+To demonstrate the principles above, we'll write an integration test that consists of the following steps:
+
+1. Import a data set into Weaviate using documented best practices ([batching](/developers/weaviate/manage-data/import) and [streaming](/developers/weaviate/manage-data/import#tip-stream-data-from-large-files))
+  a. Have vectorization enabled, so we effectively test a workflow that combines Weaviate and an external API (to control API call costs, we can limit the number of objects imported)
+2. Ensure that all specified data objects have been imported, and no extraneous ones exist
+3. Export back objects along with their vectors (useful for backups and recreating the database without hitting the vectorization provider)
```

Comment:
Weaviate has a built-in backup functionality. I don't think we should teach here how to perform a manual backup.

Also, this is not related to automated testing, which is the topic of this blog post.